src/promptflow-core/promptflow/core/_prompty_utils.py

```diff
@@ -558,22 +673,194 @@ def parse_chat(chat_str, images: List = None, valid_roles: List[str] = None):
+            return ChatInputList([Escaper.escape_roles_in_flow_input(item, escape_dict) for item in val])
+        elif isinstance(val, str):
+            for encoded_role, role in escape_dict.items():
+                val = val.replace(role, encoded_role)
+            return val
+        else:
+            return val
+
+    @staticmethod
+    def unescape_roles(val, escape_dict: dict):
+        """
+        Unescape the roles in the parsed chat messages to restore the original role names.
+        Besides the case that value is: 'some text. escaped_roles (i.e. fake uuids)'
+        We also need to handle the vision case that the content is converted to list.
+        For example:
+            [{
+                'type': 'text',
+                'text': 'some text. fake_uuid'
+            }, {
+                'type': 'image_url',
+                'image_url': {}
+            }]
+        """
+        if not escape_dict:
+            return val
+
+        if isinstance(val, str):
+            for encoded_role, role in escape_dict.items():
+                val = val.replace(encoded_role, role)
+            return val
+        elif isinstance(val, list):
+            for index, item in enumerate(val):
+                if isinstance(item, dict) and "text" in item:
+                    for encoded_role, role in escape_dict.items():
+                        val[index]["text"] = item["text"].replace(encoded_role, role)
+            return val
+        else:
+            return val
+
+    @staticmethod
+    def escape_kwargs(escape_dict: dict, _inputs_to_escape: list, **kwargs):
+        # Use escape/unescape to avoid unintended parsing of role in user inputs.
+        # There are two scenarios to consider for llm/prompt tool:
+        # 1. Prompt injection directly from flow input.
+        # 2. Prompt injection from the previous linked prompt tool, where its output becomes llm/prompt input.
+        updated_kwargs = {}
+        for k, v in kwargs.items():
+            if isinstance(v, PromptResult) and v.need_to_escape():
+                updated_kwargs[k] = v.get_escape_string()
+            elif _inputs_to_escape and k in _inputs_to_escape:
+                updated_kwargs[k] = Escaper.escape_roles_in_flow_input(v, escape_dict)
+            else:
+                updated_kwargs[k] = v
+
+        return updated_kwargs
+
+    @staticmethod
+    def build_escape_dict_from_kwargs(_inputs_to_escape: list, **kwargs):
+        prompt_result_escape_dict = Escaper.merge_escape_mapping_of_prompt_results(**kwargs)
+        flow_inputs_escape_dict = Escaper.build_flow_inputs_escape_dict(_inputs_to_escape=_inputs_to_escape, **kwargs)
+        escape_dict = {**prompt_result_escape_dict, **flow_inputs_escape_dict}
+
+        return escape_dict
+
+
+def build_messages(
+    prompt: PromptTemplate,
+    images: List = None,
+    image_detail: str = "auto",
+    **kwargs,
+):
+    inputs_to_escape = kwargs.pop(INPUTS_TO_ESCAPE_PARAM_KEY, None)
```

Comment:
It seems this PR only handles the scenario that the prompty within a flow? Since 'INPUTS_TO_ESCAPE_PARAM_KEY" is only set in executor when it is a flow input.

How about the scenario that the prompty is not within a flow?

Reply:
Fix it.